### DISCUSSION TITLE
Milvus persistant data

### DISCUSSION THREAD

**Discussion by iamkhalidbashir** [milvus-io/milvus#37424] [Q&A and General discussion]
I want to deploy Milvus on the production
- I assumed Milvus uses minio s3 storage for all its data/index/meta storage
- But it seems Milvus only stores data and indexes only on minio s3
- It stores metadata (database and collection schema) on etcd, which I assumed was just temporary key-value storage and is compacted (deleted after some time)
- I plan to use restic for periodic encrypted s3 backup rather than milvus-backup for backing up

So my question is for Milvus standalone:
Do I need to backup the etcd directory (for metadata)? or there is any sort flush etcd command to move the data to minio s3?

**Comment by liliu-z** [comment]
> I want to deploy Milvus on the production
> 
> * I assumed Milvus uses minio s3 storage for all its data/index/meta storage
> * But it seems Milvus only stores data and indexes only on minio s3
> * It stores metadata (database and collection schema) on etcd, which I assumed was just temporary key-value storage and is compacted (deleted after some time)
> * I plan to use restic for periodic encrypted s3 backup rather than milvus-backup for backing up
> 
> So my question is for Milvus standalone: Do I need to backup the etcd directory (for metadata)? or there is any sort flush etcd command to move the data to minio s3?

Milvus supports a backup tool to help this scenario https://milvus.io/docs/milvus_backup_overview.md#Milvus-Backup

**Comment by iamkhalidbashir** [comment]
Yes i am aware of it but i intend to not use it as it’s not production
ready.

For example if I use some sort of cronjob for it to backup every hour then:
- its not efficient for large data, as it doesnt have incremental backup
like restic
- also if the ectd crashes within that 1 hour all data is within that 1
hour is lost

I plan to use restic for backup

*- Khalid*

**Reply by xiaofan-luan** [reply]
It's gonna to hard to do so with cluster running because It's hard to maintain consistency between etcd and S3.

To do so, you need to:
1. Flush all data to S3
2. backup the etcd.
3. Do a S3 snapshot and Kafka (if you don't have streaming write then kafka is not necessary)

This is gonna to be very challenging to make sure all the data is consistent because meta and data is not stored in the same place. Snapshot could be a interesting feature to work with but unfortunately we don't have bandwidth to work on this.

**Comment by iamkhalidbashir** [comment]
Thanks for the tip!

May i know what is the ideal backup solution for production use of
standalone

Use backup tool to backup every 60 minutes and on server crash you loose
60mins of data ?

**Reply by xiaofan-luan** [reply]
IF we store with distributed kafka and distributed minio or S3 then no data is actually losing. 
Milvus recover from all the persistant data and that is the common way we handle crash.

**Reply by xiaofan-luan** [reply]
Even for standalone, unless the disk is crashed, milvus should be able to recover from WAL and there is no data loss.

**Comment by iamkhalidbashir** [comment]
My concern is to launch it on an ec2 server with hard ssd drive as ectd
store
Now when shutting down and running on a new server
If I just provide the files minio s3 to the new server it dosent show the
database or collections on the new server attu
But when i copy the directory of ectd, then it does show all the databases
and schema
So i assumed there is data lose when ectd directory is deleted on hard
drive crash or running a fresh server

*- Khalid*

**Reply by xiaofan-luan** [reply]
> My concern is to launch it on an ec2 server with hard ssd drive as ectd store Now when shutting down and running on a new server If I just provide the files minio s3 to the new server it dosent show the database or collections on the new server attu But when i copy the directory of ectd, then it does show all the databases and schema So i assumed there is data lose when ectd directory is deleted on hard drive crash or running a fresh server *- Khalid*
> […](#)
> On Tue, 5 Nov 2024 at 5:53 AM Xiaofan ***@***.***> wrote: Even for standalone, unless the disk is crashed, milvus should be able to recover from WAL and there is no data loss. — Reply to this email directly, view it on GitHub <[#37424 (reply in thread)](https://github.com/milvus-io/milvus/discussions/37424#discussioncomment-11148742)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AGS5WW7T6GLY4V4MV7GVRULZ7ACGNAVCNFSM6AAAAABREFOPNGVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMJUHA3TIMQ> . You are receiving this because you authored the thread.Message ID: ***@***.***>

To migrate a milvus service, you need to stop milvus, and copy data from both
1. etcd
2. kafka/pulsar/rocksmq
3. minio/S3

If you want to migrate data between two instance, try https://github.com/zilliztech/vts.

also think of using our managed service so you don't need to maintain everything and it's much easier to setup https://zilliz.com/
